pkg/lib/messaging/sender.go

```diff
@@ -190,25 +198,41 @@ func (s *Sender) SendSMSImmediately(ctx context.Context, msgType translation.Mes
 		return err
 	}
 
-	err = s.SMSSender.Send(ctx, client, *opts)
-	if err != nil {
-		otelauthgear.IntCounterAddOne(
-			ctx,
-			otelauthgear.CounterSMSRequestCount,
-			otelauthgear.WithStatusError(),
-		)
+	send := func() error {
+		err = s.SMSSender.Send(ctx, client, *opts)
+		if err != nil {
+			otelauthgear.IntCounterAddOne(
+				ctx,
+				otelauthgear.CounterSMSRequestCount,
+				otelauthgear.WithStatusError(),
+			)
 
-		// TODO: Handle expected errors https://linear.app/authgear/issue/DEV-1139
-		s.Logger.WithError(err).WithFields(logrus.Fields{
-			"phone": phone.Mask(opts.To),
-		}).Error("failed to send SMS")
-		dispatchErr := s.DispatchEventImmediatelyWithTx(ctx, &nonblocking.SMSErrorEventPayload{
-			Description: s.errorToDescription(err),
-		})
-		if dispatchErr != nil {
-			s.Logger.WithError(err).Errorf("failed to emit %v event", nonblocking.SMSError)
+			// TODO: Handle expected errors https://linear.app/authgear/issue/DEV-1139
+			s.Logger.WithError(err).WithFields(logrus.Fields{
+				"phone": phone.Mask(opts.To),
+			}).Error("failed to send SMS")
+			dispatchErr := s.DispatchEventImmediatelyWithTx(ctx, &nonblocking.SMSErrorEventPayload{
+				Description: s.errorToDescription(err),
+			})
+			if dispatchErr != nil {
+				s.Logger.WithError(err).Errorf("failed to emit %v event", nonblocking.SMSError)
+			}
+			return err
+		}
+		return nil
+	}
+
+	if isAsync {
+		go func() {
+			// Detach the deadline so that the context is not canceled along with the request.
+			ctx = context.WithoutCancel(ctx)
+			_ = send()
+		}()
```

Comment:
```golang
if isAsync {
  // Re-define ctx here so that the defer statement body captures this without-cancel ctx.
  ctx = context.WithoutCancel(ctx)
  // Pass the ctx to the goroutine explicitly.
  go func(ctx context.Context) {
      _ = send(ctx)
  }(ctx)
}
```

Reply:
If we pass ctx to the go func by parameter, why do we need redefine?

Reply:
For readability.

```golang
if isAsync {
  ctx = context.WithoutCancel(ctx)
  // Pass the ctx to the goroutine explicitly.
  go func(ctx context.Context) {
      _ = send(ctx)
  }(ctx)
}
```

vs

```golang
if isAsync {
  // Pass the ctx to the goroutine explicitly.
  go func(ctx context.Context) {
      _ = send(ctx)
  }(context.WithoutCancel(ctx))
}
```

I think the former is more readable.

Reply:
But that actually exposed the `ctx` outside the block, which can then be used incorrectly by code under it. I prefer do it thisway

```
if isAsync {
  ctxWithoutCancel := context.WithoutCancel(ctx)
  // Pass the ctx to the goroutine explicitly.
  go func(ctxWithoutCancel context.Context) {
      _ = send(ctxWithoutCancel)
  }(ctxWithoutCancel)
}
```

Reply:
Rename the argument to `ctx`. 

```golang
if isAsync {
  ctxWithoutCancel := context.WithoutCancel(ctx)
  go func(ctx context.Context) {
      _ = send(ctx)
  }(ctxWithoutCancel)
}
```